### PR TITLE
Housekeeping for 0.7

### LINKED ISSUE #6: Add screenshot generation commands to justfile
The `simulator` justfile contained some example related code, that should be moved here:
```
#----------
# Examples
#----------

# Generates a screenshot of an example
generate-example-screenshot example:
    @mkdir -p "{{screenshots_dir}}"
    # Generating {{example}} screenshot
    EG_SIMULATOR_DUMP="{{screenshots_dir}}/{{example}}.png" \
    cargo run --example {{example}}

# Generates screenshots of all examples
@generate-example-screenshots:
    #!/usr/bin/env bash
    set -e
    for example in examples/*.rs; do \
        just generate-example-screenshot $(basename "$example" .rs); \
    done
```

### LINKED ISSUE #7: Update examples with thick triangle/polyline
From https://github.com/embedded-graphics/simulator/pull/9#issuecomment-728321348, some examples need updating/improving to add a demonstration of thick polylines and thick stroked triangles, now that e-g itself supports them.

Non exhaustive suggestions so far:

- `primitives-triangles`

  > which seems to be more of a test case for different triangles than an example.
  >
  > -- https://github.com/embedded-graphics/simulator/pull/9#issuecomment-728321348

- `primitives-stroke`

### PR DESCRIPTION
**Pull request by jamwaffles**
Progress towards #8 and #5, closes #7, closes #6.

### CODE REVIEW

**Comment by jamwaffles** [comment]
Rectangle intersection moved to embedded-graphics/debug-tools#9.

**Comment by jamwaffles** [comment]
A bit of a kitchen sink PR, but I think I've covered most points in the other issues in this repo

File: eg-next/examples/primitives-fill.rs [hunk 1]
```diff
@@ -24,7 +24,7 @@ where
         .into_styled(style)
         .draw(target)?;
 
-    Rectangle::new(Point::new(96, 0), Size::new(64, 64))
+    Rectangle::new(Point::new(97, 0), Size::new(64, 64))
```
Comment (rfuest) [inline eg-next/examples/primitives-fill.rs:27]: Why is this change necessary? (Maybe related to the comment below)

File: eg-next/examples/primitives-fill.rs [hunk 2]
```diff
@@ -33,21 +33,21 @@ where
         .into_styled(style)
         .draw(target)?;
```
Comment (rfuest) [inline eg-next/examples/primitives-fill.rs:34]: I think the `stroke_off_fill_of` rectangle clips the lower left corner of the `stroke` triangle.

File: eg-next/examples/primitives-fill.rs [hunk 3]
```diff
@@ -33,21 +33,21 @@ where
...
     )
-    .translate(Point::new(96 * 4, 0))
+    .translate(Point::new(96 + 24, 97))
     .into_styled(style)
     .draw(target)
 }
 
 fn main() -> Result<(), Infallible> {
-    let mut display: SimulatorDisplay<BinaryColor> = SimulatorDisplay::new(Size::new(512, 128));
+    let mut display: SimulatorDisplay<BinaryColor> = SimulatorDisplay::new(Size::new(300, 200));
```
Comment (rfuest) [inline eg-next/examples/primitives-fill.rs:50]: Could we move all drawn shapes a bit to the bottom right, so that they look centered? Either by adding a `translated` call to all `draw_shapes` invocations or by adding an offset inside `draw_shaped`.

File: eg-next/examples/primitives-fill.rs [hunk 4]
```diff
@@ -33,21 +33,21 @@ where
...
 
     let stroke = PrimitiveStyle::with_stroke(BinaryColor::On, 1);
```
Comment (rfuest) [inline eg-next/examples/primitives-fill.rs:53]: The `stroke_off_fill_off` style behaves the same as a `PrimitiveStyle::with_fill(BinaryColor::Off)`. Would it be better to change the example to use colors to better differentiate the styles?

File: eg-next/examples/primitives-triangles.rs [hunk 5]
```diff
@@ -1,60 +1,96 @@
...
-    Triangle::new(Point::new(5, 0), Point::new(30, 64), Point::new(64, 0))
-        .translate(Point::new(64 + PAD, 0))
-        .into_styled(style)
+    // Inside thick stroke, no fill
+    base_triangle
+        .into_styled(
+            PrimitiveStyleBuilder::new()
+                .stroke_alignment(StrokeAlignment::Inside)
+                .stroke_width(10)
+                .stroke_color(Rgb888::CSS_SALMON)
```
Comment (rfuest) [inline eg-next/examples/primitives-triangles.rs:42]: I try to follow this order in the examples:
```suggestion
            PrimitiveStyleBuilder::new()
                .stroke_color(Rgb888::CSS_SALMON)
                .stroke_width(10)
                .stroke_alignment(StrokeAlignment::Inside)
```
The reason to use this order is that the first two calls are in the same order as the arguments in `PrimitiveStyle::with_stroke`. And the alignment is an advanced option that should follow the basic settings IMO.

File: eg-next/examples/primitives-triangles.rs [hunk 6]
```diff
@@ -1,60 +1,96 @@
...
-        .translate(Point::new((64 + PAD) * 2, 0))
-        .into_styled(style)
+    // Center stroke alignment with fill
+    flipped_triangle
+        .translate(Point::new(offset, 0))
+        .into_styled(
+            PrimitiveStyleBuilder::new()
+                .stroke_width(10)
+                .stroke_color(Rgb888::CSS_AQUAMARINE)
+                .fill_color(Rgb888::CSS_CADET_BLUE)
```
Comment (rfuest) [inline eg-next/examples/primitives-triangles.rs:54]:
```suggestion
                .stroke_color(Rgb888::CSS_AQUAMARINE)
                .stroke_width(10)
                .fill_color(Rgb888::CSS_CADET_BLUE)
```

File: eg-next/examples/primitives-triangles.rs [hunk 7]
```diff
@@ -1,60 +1,96 @@
...
-        .into_styled(style)
+    // Outside stroke alignment with fill
+    base_triangle
+        .translate(Point::new(offset * 2, 0))
+        .into_styled(
+            PrimitiveStyleBuilder::new()
+                .stroke_alignment(StrokeAlignment::Outside)
+                .stroke_width(9)
+                .stroke_color(Rgb888::CSS_FIRE_BRICK)
+                .fill_color(Rgb888::CSS_WHITE_SMOKE)
```
Comment (rfuest) [inline eg-next/examples/primitives-triangles.rs:67]:
```suggestion
                .stroke_color(Rgb888::CSS_FIRE_BRICK)
                .stroke_width(9)
                .stroke_alignment(StrokeAlignment::Outside)
                .fill_color(Rgb888::CSS_WHITE_SMOKE)
```

File: eg-next/examples/text-alignment.rs [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+        .build();
+
+    Text::with_text_style(
+        "Left aligned text, origin top left",
+        bounding_box.top_left,
+        character_style,
+        TextStyleBuilder::new()
+            .alignment(Alignment::Left)
+            .baseline(Baseline::Top)
+            .build(),
```
Comment (rfuest) [inline eg-next/examples/text-alignment.rs:30]: I would suggest to define all styles before drawing any text.

This is a general suggestions for all inline style builder calls. I've seen multiple examples where people have written wrappers around text drawing or other draw functions because our calls seem to be to complicated to use every time. But IMO many of these problems could have been solved by creating styles ahead of time.

File: eg-next/examples/text-alignment.rs [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+        character_style,
+        TextStyleBuilder::new()
+            .alignment(Alignment::Right)
+            .baseline(Baseline::Bottom)
+            .build(),
+    )
+    .draw(&mut display)?;
+
+    let output_settings = OutputSettingsBuilder::new().scale(2).build();
+    Window::new("Fonts", &output_settings).show_static(&display);
```
Comment (rfuest) [inline eg-next/examples/text-alignment.rs:57]:
```suggestion
    Window::new("Text alignment", &output_settings).show_static(&display);
```

File: eg-next/examples/text-styles.rs [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    )
+    .draw(&mut display)?;
+    let position = Text::new(", ", position, normal).draw(&mut display)?;
+    let position = Text::new("highlighted", position, background).draw(&mut display)?;
+    let position = Text::new(" and ", position, normal).draw(&mut display)?;
+    let position = Text::new("italic", position, italic).draw(&mut display)?;
+    Text::new(" text!", position, normal).draw(&mut display)?;
+
+    let output_settings = OutputSettingsBuilder::new().scale(2).build();
+    Window::new("Fonts", &output_settings).show_static(&display);
```
Comment (rfuest) [inline eg-next/examples/text-styles.rs:68]: Nice example!

```suggestion
    Window::new("Text styles", &output_settings).show_static(&display);
```

**Review by rfuest — CHANGES_REQUESTED** [review]
First round of reviews

**Comment by rfuest** [comment]
> Closes #8, closes #7, closes #6, closes #5

#8 still need some work after this PR is done.

And this PR is a good start towards #5, but could also use some work. Should we also add licensing information to the readme?

**Comment by rfuest** [comment]
* Should we move `simulator-input-handling` to the examples directory in the `simulator` repo?
* Should we rename `line-thickness` to `primitives-line-thickness` or also move it `debug-tools`?
* Can we set the initial progress value in `demo-progress` to something != 0? This would make the screenshot more interesting.
* `text-custom-font` will need to be fixed after the next beta release.

**Comment by jamwaffles** [comment]
To address your points in order:

1. Agreed, that example really doesn't make sense here does it. Moved in embedded-graphics/simulator#26
2. I've deleted it, as we already have the `line` debug tool which does the same thing AFAICS.
3. Set to >0 now, looks better indeed.
4. :+1: 

As for the licenses, I can use the same structure as the e-g repo. Let me do that quick...

File: eg-next/examples/primitives-fill.rs (hunk 1 above)
Comment (jamwaffles) [reply eg-next/examples/primitives-fill.rs]: Well spotted, yes that was to fix some clipping. I've now fixed the new clipping this changed caused so we should be all good now.

File: eg-next/examples/primitives-fill.rs (hunk 3 above)
Comment (jamwaffles) [reply eg-next/examples/primitives-fill.rs]: Done. I've also added some spacing too. The previous `BinaryColor` rendering was actually pretty centered, just the invisible final shape was moving the "visual center" over a bit.

File: eg-next/examples/text-alignment.rs (hunk 8 above)
Comment (jamwaffles) [reply eg-next/examples/text-alignment.rs]: I think I've caught every instance of `TextStyleBuilder` and `PrimitiveStyleBuilder` and moved them to their own variables. Let me know if I missed one.

**Comment by jamwaffles** [comment]
Licenses are added, MIT or Apache 2.0 the same as e-g. I also added them to the `Cargo.toml` of each folder which isn't really necessary, but can't hurt.

**Comment by rfuest** [comment]
Some final comments:
* I would prefer to keep the analog prefix for the clock demo, because I'm considering to add a `eg-seven-segment` based digital clock example to this repo: https://github.com/embedded-graphics/eg-seven-segment/blob/master/examples/digital-clock.rs
* IMO `primitives-ellipse` and `primitives-rounded-rectangles` are debug tools. Should we remove them? Both primitives are used in other examples.
* The comment in `text-fonts` says: "Demonstrate the available builtin fonts." We should either change the comment to state that this is a subset of available fonts and a full list is included in the docs or add all fonts to the example. I guess the first solution would be easier to implement and maintain in the future.
* `text-extended-characters`: The comment needs to be updated, because all builtin fonts support the characters in this example

**Comment by jamwaffles** [comment]
Changes addressed :)

File: LICENSE-MIT [hunk 11]
```diff
@@ -0,0 +1,25 @@
+Copyright (c) 2020 James Waples
+
```
Comment (rfuest) [inline LICENSE-MIT:2]: Other creates in the Rust ecosystem (including the main rust repo) have removed this line from the MIT license. I think the TLDR was that each author retains the copyright to their contribution and that a global copyright is invalid anyway.

```suggestion
```

**Review by rfuest — CHANGES_REQUESTED** [review]
Looks good, only one small issue left.

File: LICENSE-MIT (hunk 11 above)
Comment (jamwaffles) [reply LICENSE-MIT]: Makes sense. I've removed it from the license file.

**Comment by rfuest** [comment]
Should we "fork" the eg-0.7 examples directory in this PR? Because the `Cargo.toml` in `eg-next` was supposed to always use git dependencies.

**Comment by jamwaffles** [comment]
Ah, no, I think we should do that when 0.7 is released. Technically there's no proper release yet, so keeping `eg-next` as is for the short term makes sense to me. I've reverted the dependency changes back to the git/patches for this PR.

**Comment by rfuest** [comment]
OK. But I have to admit that finally seeing the `Cargo.toml` without any git dependencies looked really good.

**Review by rfuest — APPROVED** [review]

**Comment by jamwaffles** [comment]
Thanks for the reviews! These examples look really good now.

> But I have to admit that finally seeing the Cargo.toml without any git dependencies looked really good.

It did look spectacular! The quicker we can get the migration guide done the quicker we can release non-beta and go back to non-Git `Cargo.toml` utopia ;)
